plugins/processors/starlark/starlark.go

```diff
@@ -0,0 +1,217 @@
+
+	program, err := s.sourceProgram(builtins)
+	if err != nil {
+		return err
+	}
+
+	// Execute source
+	globals, err := program.Init(s.thread, builtins)
+	if err != nil {
+		return err
+	}
+
+	// Freeze the global state.  This prevents modifications to the processor
+	// state and prevents scripts from containing errors storing tracking
+	// metrics.  Tasks that require global state will not be possible due to
+	// this, so maybe we should relax this in the future.
+	globals.Freeze()
+
+	// The source should define an apply function.
+	apply := globals["apply"]
+
+	if apply == nil {
+		return errors.New("apply is not defined")
+	}
+
+	var ok bool
+	if s.applyFunc, ok = apply.(*starlark.Function); !ok {
+		return errors.New("apply is not a function")
+	}
+
+	if s.applyFunc.NumParams() != 1 {
+		return errors.New("apply function must take one parameter")
+	}
+
+	// Reusing the same metric wrapper to skip an allocation.  This will cause
+	// any saved references point to the new metric, but due to freezing the
+	// globals none should exist.
+	s.args = make(starlark.Tuple, 1)
+	s.args[0] = &Metric{}
+
+	// Preallocate a slice for return values.
+	s.results = make([]telegraf.Metric, 0, 10)
+
+	return nil
+}
+
+func (s *Starlark) sourceProgram(builtins starlark.StringDict) (*starlark.Program, error) {
+	if s.Source != "" {
+		_, program, err := starlark.SourceProgram("processor.starlark", s.Source, builtins.Has)
+		return program, err
+	}
+	_, program, err := starlark.SourceProgram(s.Script, nil, builtins.Has)
+	return program, err
+}
+
+func (s *Starlark) SampleConfig() string {
+	return sampleConfig
+}
+
+func (s *Starlark) Description() string {
+	return description
+}
+
+func (s *Starlark) Start(acc telegraf.Accumulator) error {
+	return nil
+}
+
+func (s *Starlark) Add(metric telegraf.Metric, acc telegraf.Accumulator) {
+	s.args[0].(*Metric).Wrap(metric)
+
+	rv, err := starlark.Call(s.thread, s.applyFunc, s.args, nil)
+	if err != nil {
+		if err, ok := err.(*starlark.EvalError); ok {
+			for _, line := range strings.Split(err.Backtrace(), "\n") {
+				s.Log.Error(line)
+			}
+		} else {
+			s.Log.Errorf("Error calling Starlark: %v", err)
+		}
+		metric.Reject()
+		return
+	}
+
+	switch rv := rv.(type) {
+	case *starlark.List:
+		iter := rv.Iterate()
+		defer iter.Done()
+		var v starlark.Value
+		for iter.Next(&v) {
+			switch v := v.(type) {
+			case *Metric:
+				m := v.Unwrap()
+				if containsMetric(s.results, m) {
+					s.Log.Errorf("Duplicate metric reference detected")
+					continue
+				}
+				s.results = append(s.results, m)
+				acc.AddMetric(m)
+			case starlark.Value:
+				s.Log.Errorf("Invalid type returned in list: %s", v.Type())
+			}
+		}
+
+		// If the script didn't return the original metrics, mark it as
+		// successfully handled.
+		if !containsMetric(s.results, metric) {
+			metric.Accept()
+		}
+
+		// clear results
+		for i := range s.results {
+			s.results[i] = nil
+		}
```

Comment:
I think so, what are you worried about?

Reply:
Not sure if it would actually reuse the memory addresses on append, which I'm guessing is the desired behavior. Just unfamiliar with this. (I'm assuming you're trying to leave the capacity and reset the length)